specification/dartLangSpec.tex

```diff
@@ -14107,293 +14131,569 @@ \section{Libraries and Scripts}
+    When a name $n$ has been eliminated for this reason and a usage
+    \code{$p_i$.$n$} is encountered in $L$,
+    it is recommended that the error message mentions the conflict,
+    rather than just reporting that $n$ does not exist.%
+  }
+  % We need 'not the same library', l != j is not enough: See comment on the
+  % corresponding case for step two below (search 'not the same library').
+\end{itemize}
 
 \LMHash{}%
-After a call succeeds, the name $p$ is mapped to a non-deferred prefix object as described below.
-In addition, the prefix object also supports the \code{loadLibrary} method, and so it is possible to call \code{loadLibrary} again.
-If a call fails, nothing happens, and one again has the option to call \code{loadLibrary} again.
-Whether a repeated call to \code{loadLibrary} succeeds will vary as described below.
+Let \NamespaceName{\metavar{prefix},i} be the namespace
+$\NamespaceName{\metavar{narrowed},1} \cup \ldots
+\NamespaceName{\metavar{narrowed},k}$.
+Then \NamespaceName{\metavar{one},i} is the namespace that
+maps $p_i$ to a \Index{prefix object},
+which is a mapping that associates names with entities declared in
+\List{L'}{1}{k} as follows:
 
-\LMHash{}%
-The effect of a repeated call to \code{$p$.loadLibrary} is as follows:
 \begin{itemize}
 \item
-If another call to \code{$p$.loadLibrary} has already succeeded, the repeated call also succeeds.
-Otherwise,
+  For every top level function $f$ named \id{} in
+  \NamespaceName{\metavar{prefix}, i},
+  \id{} is mapped to a function with the same name and signature as $f$.
 \item
-If another call to \code{$p$.loadLibrary} has failed:
-\begin{itemize}
+  For every top level getter $g$ named \id{} in
+  \NamespaceName{\metavar{prefix}, i},
+  \id{} is mapped to a getter with the same return type as $g$.
 \item
-If the failure is due to a compilation error, the repeated call fails for the same reason.
+  For every top level setter $s$ named \code{\id=} in
+  \NamespaceName{\metavar{prefix}, i},
+  \code{\id=} is mapped to a setter with the same parameter type as $s$.
 \item
-If the failure is due to other causes, the repeated call behaves as if no previous call had been made.
-\end{itemize}
+  For every type $T$ named \id{} in
+  \NamespaceName{\metavar{prefix}, i},
+  \id{} is mapped to the type $T$.
 \end{itemize}
 
-\commentary{
-In other words, one can retry a deferred load after a network failure
-or because a file is absent,
-but once one finds some content and loads it, one can no longer reload.
+\commentary{%
+A prefix object is not a Dart object, it is merely a device which is
+used to manage names of the form \synt{qualified} and their types
+during static analysis,
+and a mapping that is used to access the corresponding entities
+at run time.
+Consequently, any attempt to use a prefix object on its own
+is a compile-time error.
 
-We do not specify which object the future returned resolves to.
+Note that if $l$ and $q$ are such that
```

Comment:
If $I_l$ and $I_q$ have the same prefix $p$ then `NS_{one,l}` and `NS_{one,q}` are computed based on $p$ (when we've found $p$, $l$ and $q$ are ignored), so those namespaces are indeed the same, and the same as their union.

Reply:
ACK. Introduing a namespace for a prefix with one entry mapping the prefix to another namespace is a little subtle. Might want to mention it explicitly up-front instead of just doing it deep in the algorithm.

Reply:
I'll handle the placement of the definition of prefix namespaces in #1161.